xla/service/gpu/gemm_rewriter.cc

```diff
@@ -858,33 +865,176 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
+        maximum = op;
+        break;
+      }
+    }
+    int64_t fwd_gemm_num_rows = fwd_gemm->shape().dimensions(0);
+    if (maximum == nullptr ||
+        !SupportsEpilogueFusion(fwd_gemm->shape().element_type()) ||
+        (fwd_gemm_num_rows % 128 != 0)) {
+      // cublasLt requires that CUBLASLT_MATMUL_DESC_EPILOGUE_AUX_LD Must be
+      // divisible by 128 and be no less than the number of rows in the output
+      // matrix.
+      VLOG(1) << "Possible GEMM with ReLU epilogue is "
+              << "not fused into Custom Call.";
+      return absl::OkStatus();
+    }
+
+    // Rewrite Fwd Relu with ReluAux
+    TF_ASSIGN_OR_RETURN(auto gpu_config,
+                        fwd_gemm->backend_config<GpuBackendConfig>());
+    GemmBackendConfig &config = *gpu_config.mutable_gemm_backend_config();
+    if (config.beta() != 0.0) {
+      // Matrix bias add is not supported by BIAS_RELU_AUX epilogue.
+      return absl::OkStatus();
+    }
+
+    if (config.epilogue() == GemmBackendConfig::DEFAULT) {
+      config.set_epilogue(GemmBackendConfig::RELU_AUX);
+    } else if (config.epilogue() == GemmBackendConfig::BIAS) {
+      config.set_epilogue(GemmBackendConfig::BIAS_RELU_AUX);
+    } else {
+      return absl::OkStatus();
+    }
+
+    Shape mask_shape = ShapeUtil::MakeShape(
+        PrimitiveType::U8,
+        {ShapeUtil::ElementsIn(fwd_gemm->shape()) / CHAR_BIT});
+    std::unique_ptr<HloInstruction> output = fwd_gemm->CloneWithNewShape(
+        ShapeUtil::MakeTupleShape({fwd_gemm->shape(), mask_shape}));
+    TF_RETURN_IF_ERROR(output->set_backend_config(gpu_config));
+    TF_RETURN_IF_ERROR(SetName(output->GetModule(), output.get()));
+    HloInstruction *tuple_output =
+        fwd_gemm->parent()->AddInstruction(std::move(output));
+    TF_RETURN_IF_ERROR(ReplaceWithNewInstruction(
+        maximum, HloInstruction::CreateGetTupleElement(tuple_output, 0)));
+
+    // Insert DRELU in backward matmul
+    std::vector<HloInstruction *> operands(bwd_gemm->operands().begin(),
+                                           bwd_gemm->operands().end());
+    operands.insert(operands.end(), fwd_gemm->AddInstruction(
+                                        HloInstruction::CreateGetTupleElement(
+                                            tuple_output, 1)));
+
+    HloInstruction *new_bwd_custom_call = bwd_gemm->AddInstruction(
+        bwd_gemm->CloneWithNewOperands(bwd_gemm->shape(), operands));
+
+    TF_ASSIGN_OR_RETURN(auto bwd_gpu_backend_config,
+                        bwd_gemm->backend_config<GpuBackendConfig>());
+    GemmBackendConfig &bwd_config =
+        *bwd_gpu_backend_config.mutable_gemm_backend_config();
+    bwd_config.set_epilogue(GemmBackendConfig::D_RELU);
+    TF_RETURN_IF_ERROR(
+        new_bwd_custom_call->set_backend_config(bwd_gpu_backend_config));
+
+    return ReplaceInstruction(instr, new_bwd_custom_call);
+  }
+
+  // Attempt to match the backward pass of a vector bias addition only when
+  // there is a ReLU.
```

Comment:
Why only when there is a ReLU? If we just didn't implement the case without a Relu, I would phrase this as:

> Attempt to match the gradient of a vector bias addition, which is a reduction. Currently we only match this when there is a Relu gradient before the bias addition gradient.